### PR TITLE
feat(index): introduce setIndexUiState

### PR DESCRIPTION
**Pull request by Haroenv**
**Summary**

This function is visible in useInstantSearch, so it has potential to supplant the usage of `helper.setState().search` in the future.

**Result**

- adds index.setIndexUiState
- use index.setUiState in useInstantSearch
- allow IndexWidget type to be generic

### CODE REVIEW

**Review by FabienMotte — APPROVED** [review]

File: packages/instantsearch.js/src/widgets/index/__tests__/index-test.ts [hunk 1]
```diff
@@ -2894,6 +2902,64 @@ See documentation: https://www.algolia.com/doc/api-reference/widgets/index-widge
     });
   });
 
+  describe('setIndexUiState', () => {
+    it('updates main ui state with object', () => {
```
Comment (sarahdayan) [inline packages/instantsearch.js/src/widgets/index/__tests__/index-test.ts:2906]:
```suggestion
    it('updates main UI state with an object', () => {
```

File: packages/instantsearch.js/src/widgets/index/__tests__/index-test.ts [hunk 2]
```diff
@@ -2894,6 +2902,64 @@ See documentation: https://www.algolia.com/doc/api-reference/widgets/index-widge
...
+
+      expect(instantSearchInstance.getUiState()).toEqual({
+        root: {},
+        indexName: {
+          query: 'iphone',
+        },
+      });
+    });
+
+    it('updates main ui state with function', () => {
```
Comment (sarahdayan) [inline packages/instantsearch.js/src/widgets/index/__tests__/index-test.ts:2929]:
```suggestion
    it('updates main UI state with a function', () => {
```

File: packages/instantsearch.js/src/widgets/index/__tests__/index-test.ts [hunk 3]
```diff
@@ -2894,6 +2902,64 @@ See documentation: https://www.algolia.com/doc/api-reference/widgets/index-widge
...
+      expect(instantSearchInstance.getUiState()).toEqual({
+        root: {},
+        indexName: {
+          page: 2,
+        },
+      });
+
+      instance.setIndexUiState((uiState) => ({
+        page: (uiState.page || 1) + 1,
+      }));
```
Comment (sarahdayan) [inline packages/instantsearch.js/src/widgets/index/__tests__/index-test.ts:2952]: Is the user responsible for merging with the existing UI state if they want to preserve other keys, or do we merge?

**Review by sarahdayan — APPROVED** [review]

File: packages/instantsearch.js/src/widgets/index/__tests__/index-test.ts (hunk 3 above)
Comment (Haroenv) [reply packages/instantsearch.js/src/widgets/index/__tests__/index-test.ts:2952]: just like the main setUiState, the user is merging

**Pull request merged**